crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,985 @@
+            // Operation::trunc
+            // jumps, ass, nop => do not evaluate
+            // Operation::ass => todo!(),
+            // Operation::nop => todo!(),
+        }
+        NodeEval::Idx(self.idx)
+    }
+
+    // Simplifies trivial Phi instructions by returning:
+    // None, if the instruction is unreachable or in the root block and can be safely deleted
+    // Some(id), if the instruction can be replaced by the node id
+    // Some(ins_id), if the instruction is not trivial
+    pub fn simplify_phi(
+        ins_id: arena::Index,
+        phi_arguments: &[(arena::Index, arena::Index)],
+    ) -> Option<arena::Index> {
+        let mut same = None;
+        for op in phi_arguments {
+            if Some(op.0) == same || op.0 == ins_id {
+                continue;
+            }
+            if same.is_some() {
+                //no simplification
+                return Some(ins_id);
+            }
+
+            same = Some(op.0);
+        }
+        //if same.is_none()  => unreachable phi or in root block, can be replaced by ins.lhs (i.e the root) then.
+        same
+    }
+
+    pub fn standard_form(&mut self) {
+        match self.operator {
+            //convert greater into less
+            Operation::ugt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ult
+            }
+            Operation::uge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::ule
+            }
+            Operation::sgt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::slt
+            }
+            Operation::sge => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::sle
+            }
+            Operation::gt => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lt
+            }
+            Operation::gte => {
+                std::mem::swap(&mut self.rhs, &mut self.lhs);
+                self.operator = Operation::lte
+            }
+            //TODO replace a<b with a<=b+1, but beware of edge cases!
+            Operation::eq_gate => {
+                if self.rhs == self.lhs {
+                    self.rhs = self.idx;
+                    self.is_deleted = true;
+                }
+            }
+            _ => (),
+        }
+        if is_commutative(self.operator) && self.rhs < self.lhs {
+            std::mem::swap(&mut self.rhs, &mut self.lhs);
+        }
+    }
+}
+
+//adpated from LLVM IR
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum Operation {
+    add,  //(+)
+    sadd, //(+) safe addtion
+    sub,  //(-)
+    ssub, //(-) safe substraction
+    mul,  //(*)
+    smul, //(*) safe multiplication
+    udiv, //(/) unsigned division
+    sdiv, //(/) signed division
```

Comment:
This a good catch!! In fact I realise that in some places I was using `sadd ` as signed addition instead of safe add...
The point is that since none of them is implemented, I don't know yet how this will turn out. But if we need both, it should definitively be `safe_add ` for safe and `sadd ` for signed.

Reply:
for now I use explicitly safe_add, although neither safe or signed addition is implemented.